File: src/main/java/com/timepath/swing/ThemeSelector.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.timepath.swing;

import com.timepath.launcher.util.Utils;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @author TimePath
 */
@SuppressWarnings("serial")
public class ThemeSelector extends JComboBox<String> {

    private static final Logger LOG = Logger.getLogger(ThemeSelector.class.getName());

    public ThemeSelector() {
        final DefaultComboBoxModel<String> model = new DefaultComboBoxModel<>();
        setModel(model);
        String currentLafClass = UIManager.getLookAndFeel().getClass().getName();
        for (UIManager.LookAndFeelInfo lafInfo : UIManager.getInstalledLookAndFeels()) {
            try {
                Class.forName(lafInfo.getClassName());
            } catch (ClassNotFoundException ignored) {
                continue; // Registered but not found on classpath
            }
            String name = lafInfo.getName();
            model.addElement(name);
            if (lafInfo.getClassName().equals(currentLafClass)) {
                model.setSelectedItem(name);
            }
        }
        addActionListener(new ActionListener() {
            @Override
            public void actionPerformed(ActionEvent evt) {
                String target = (String) model.getSelectedItem();
                for (UIManager.LookAndFeelInfo info : UIManager.getInstalledLookAndFeels()) {
                    if (target.equals(info.getName())) {
                        LOG.log(Level.INFO, "Setting L&F: {0}", info.getClassName());
                        try {
                            String usrTheme = info.getClassName();
                            UIManager.setLookAndFeel(usrTheme);
                            for (Window w : Window.getWindows()) { // TODO: Instrumentation to access detached components
                                SwingUtilities.updateComponentTreeUI(w);
                            }
                            Utils.SETTINGS.put("laf", usrTheme);
                            return;
                        } catch (InstantiationException | IllegalAccessException | UnsupportedLookAndFeelException e) {
                            LOG.log(Level.SEVERE, null, e);
                        } catch (ClassNotFoundException e) {
                            LOG.log(Level.WARNING, "Unable to load user L&F", e);
                        }
                    }
                }
            }
        });
    }
}

File: src/main/kotlin/com/timepath/swing/ThemeSelector.kt
package com.timepath.swing

import com.timepath.launcher.LauncherUtils
import java.awt.Window
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.util.logging.Level
import java.util.logging.Logger
import javax.swing.*

/**
 * @author TimePath
 */
SuppressWarnings("serial")
public class ThemeSelector : JComboBox<String>() {

    init {
        val model = DefaultComboBoxModel<String>()
        setModel(model)
        val currentLafClass = UIManager.getLookAndFeel().javaClass.getName()
        for (lafInfo in UIManager.getInstalledLookAndFeels()) {
            try {
                Class.forName(lafInfo.getClassName())
            } catch (ignored: ClassNotFoundException) {
                continue // Registered but not found on classpath
            }

            val name = lafInfo.getName()
            model.addElement(name)
            if (lafInfo.getClassName() == currentLafClass) {
                model.setSelectedItem(name)
            }
        }
        addActionListener(object : ActionListener {
            override fun actionPerformed(evt: ActionEvent) {
                val target = model.getSelectedItem() as String
                for (info in UIManager.getInstalledLookAndFeels()) {
                    if (target == info.getName()) {
                        LOG.log(Level.INFO, "Setting L&F: {0}", info.getClassName())
                        try {
                            val usrTheme = info.getClassName()
                            UIManager.setLookAndFeel(usrTheme)
                            for (w in Window.getWindows()) {
                                // TODO: Instrumentation to access detached components
                                SwingUtilities.updateComponentTreeUI(w)
                            }
                            LauncherUtils.SETTINGS.put("laf", usrTheme)
                            return
                        } catch (e: InstantiationException) {
                            LOG.log(Level.SEVERE, null, e)
                        } catch (e: IllegalAccessException) {
                            LOG.log(Level.SEVERE, null, e)
                        } catch (e: UnsupportedLookAndFeelException) {
                            LOG.log(Level.SEVERE, null, e)
                        } catch (e: ClassNotFoundException) {
                            LOG.log(Level.WARNING, "Unable to load user L&F", e)
                        }

                    }
                }
            }
        })
    }

    companion object {

        private val LOG = Logger.getLogger(javaClass<ThemeSelector>().getName())
    }
}
